core/unit_test/TestScan.hpp

```diff
@@ -170,11 +170,108 @@ struct TestScan {
   }
 };
 
+template <class Device, class T>
+struct TestSmallSizeTypeScan {
+  using execution_space = Device;
+  using value_type      = T;
+
+  TestSmallSizeTypeScan(const size_t N, bool use_total) : x("in", N) {
+    using exec_policy = Kokkos::RangePolicy<execution_space>;
+
+    Kokkos::View<int, Device> errors_a("Errors");
+    Kokkos::deep_copy(errors_a, 0);
+    errors = errors_a;
+
+    Kokkos::deep_copy(x, 1);
+
+    T res = T(0);
+
+    if (use_total)
+      Kokkos::parallel_scan(exec_policy(0, N), *this, res);
+    else
+      Kokkos::parallel_scan(exec_policy(0, N), *this);
+    Kokkos::fence();
+
+    check_error(res, N, use_total);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(int i, value_type& update, bool final) const {
+    const auto val = x(i);
+    if (final) {
+      x(i) = update;
+
+      // We should have a nice count from 0 to 1...
+      if (update != static_cast<value_type>(i)) {
+        int fail = errors()++;
+
+        // Limit the amount of output
+        if (fail < 20) {
+          KOKKOS_IMPL_DO_NOT_USE_PRINTF(
+              "TestSmallSizeTypeScan(%d) = %ld != %ld\n", i,
+              static_cast<long>(update), static_cast<long>(i));
+        }
+      }
+    }
+    update += val;
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void init(value_type& update) const { update = 0; }
+
+  KOKKOS_INLINE_FUNCTION
+  void join(value_type& update, const value_type& input) const {
+    update += input;
+  }
+
+  void check_error(T res, std::size_t N, bool use_total) {
+    int total_errors;
+    Kokkos::deep_copy(total_errors, errors);
+    ASSERT_EQ(total_errors, 0);
+
+    if (use_total) {
+      ASSERT_EQ(res, static_cast<T>(N));
+    }
+  }
+
+  Kokkos::View<int, Device, Kokkos::MemoryTraits<Kokkos::Atomic> > errors;
+  Kokkos::View<value_type*, Device> x;
+};
+
+}  // namespace
+
 TEST(TEST_CATEGORY, scan) {
   TestScan<TEST_EXECSPACE>::test_range(1, 1000);
   TestScan<TEST_EXECSPACE>(0);
   TestScan<TEST_EXECSPACE>(100000);
   TestScan<TEST_EXECSPACE>(10000000);
   TEST_EXECSPACE().fence();
 }
-}  // namespace Test
+
+TEST(TEST_CATEGORY, small_size_scan) {
+  TestSmallSizeTypeScan<TEST_EXECSPACE, std::int8_t>(0, false);
```

Comment:
It's not intuitive to me to understand what the second parameter could do just by seeing these lines. Would you mind introducing another named variable to make this clear?

Reply:
removed this value, but did similarly with the new ImbalanceSz param